File: Tree/Leet226_invertBinaryTree.js
/*********************************************************************** 
Source LeetCode
226. Invert Binary Tree
https://leetcode.com/problems/invert-binary-tree/description/
1st 2024-06-16

Given the root of a binary tree, invert the tree, and return its root.

Example 1:
Input: root = [4,2,7,1,3,6,9]
Output: [4,7,2,9,6,3,1]

Example 2:
Input: root = [2,1,3]
Output: [2,3,1]

 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }

 * @param {TreeNode} root
 * @return {TreeNode}
************************************************************************/

// 1st Attempt
// LOGIC: recursion, go through each tree and swap
// Time : O(n)  |  Space: O(n)
var invertTree = function (root) {
  if (root == null) return null;

  let temp = root.left;
  root.left = root.right;
  root.right = temp;

  if (root.left != null) invertTree(root.left);
  if (root.right != null) invertTree(root.right);

  return root;
};

File: Tree/Leet226_invert_binary_tree.py
"""
Source LeetCode
226. Invert Binary Tree
Date: 2024-09-10
https://leetcode.com/problems/invert-binary-tree/description/

Given the root of a binary tree, invert the tree, and return its root.

Example 1:
Input: root = [4,2,7,1,3,6,9]
Output: [4,7,2,9,6,3,1]

Example 2:
Input: root = [2,1,3]
Output: [2,3,1]

Constraints:
    The number of nodes in the tree is in the range [0, 100].
    -100 <= Node.val <= 100

:type root: TreeNode
:rtype: TreeNode

# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
"""

# 1st Attempt
# LOGIC: find the position using binary search
# Time : O(logn)  |  Space: O(1)
class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """

        if root is None:
            return None
        temp = root.left
        root.left = root.right
        root.right = temp
        self.invertTree(root.left)
        self.invertTree(root.right)
        return root